### PR TITLE
[@vercel/edge] Support overriding request headers

### PR DESCRIPTION
**Pull request by nuta**
This PR adds a feature in middleware to add, modify, or delete request headers. This feature is quite useful to pass data from middleware to Serverless/Edge API routes.

### New APIs

Adds a new option `request.headers` to the `MiddlewareResponseInit` parameter in `NextResponse.next()` and `NextResponse.rewrite()`. It's a [`Header`](https://developer.mozilla.org/en-US/docs/Web/API/Headers) object holding *all* request headers. Specifically:

```ts
interface ExtraResponseInit extends ResponseInit {
  request?: {
    headers?: Headers
  }
}
```

### Example

```ts
// api/hello.ts
export default (req, res) => {
  const valueFromMiddleware = req.headers['x-hello-from-middleware']
  return res.send(valueFromMiddleware)
}

// middleware.ts
import { next } from '@vercel/edge'

export default function middleware(request: NextRequest) {
  // Clone request headers
  const headers = new Headers(request.headers);
  // Add a new request header
  headers.set('x-hello-from-middleware', 'foo');
  // Delete a request header from the client
  headers.delete('x-from-client');

  return next({
    request: {
      headers
    }
  });
}
```

### New middleware headers

- `x-middleware-override-headers`: A comma separated list of *all* request header names. Headers not listed will be deleted.
- `x-middleware-request-<name>`: A new value for the header `<name>`.

### Related Issues

- Next.js' implementation: https://github.com/vercel/next.js/pull/41380

#### Tests

- [ ] The code changed/added as part of this PR has been covered with tests
- [ ] All tests pass locally with `yarn test-unit`

#### Code Review

- [ ] This PR has a concise title and thorough description useful to a reviewer
- [ ] Issue from task tracker has a link to this PR

### CODE REVIEW

File: packages/edge/src/middleware-helpers.ts [hunk 1]
```diff
@@ -1,9 +1,39 @@
+export interface ModifiedRequest {
+  /**
+   * If this is set, the request headers will be overridden with this value.
+   */
```
Comment (Schniz) [inline packages/edge/src/middleware-helpers.ts]: Let's add the examples in the docs that show that the original headers are most likely to be cloned and passed modified (if I got this correctly):

```suggestion
  /**
   * If set, overwrites the incoming headers to the origin request.
   *
   * This is useful when you want to pass data between a Middleware and a
   * Serverless or Edge Function.
   *
   * @example
   * <caption>Add a `x-user-id` header and remove the `Authorization` header</caption>
   *
   * ```ts
   * import { rewrite } from '@vercel/edge';
   * export default async function middleware(request: Request): Promise<Response> {
   *   const newHeaders = new Headers(request.headers);
   *   newHeaders.set('x-user-id', 'user_123');
   *   newHeaders.delete('authorization');
   *   return rewrite(request.url, {
   *     request: { headers: newHeaders }
   *   })
   * }
   * ```
   */
```

File: packages/edge/src/middleware-helpers.ts [hunk 2]
```diff
@@ -1,9 +1,39 @@
...
+}
+
 export interface ExtraResponseInit extends Omit<ResponseInit, 'headers'> {
   /**
    * These headers will be sent to the user response
    * along with the response headers from the origin.
    */
   headers?: HeadersInit;
+  /**
+   * These fields will override the request from clients.
```
Comment (Schniz) [inline packages/edge/src/middleware-helpers.ts]:
```suggestion
   * Fields to rewrite for the upstream request
```

<picture data-single-emoji=":maybe-test:" title=":maybe-test:">:maybe-test:</picture> ? I am not good with lingo maybe @molebox can give us a hand? <picture data-single-emoji=":eyeses:" title=":eyeses:">:eyeses:</picture>

**Review by Schniz — APPROVED** [review]
<picture data-single-emoji=":chefkiss:" title=":chefkiss:">:chefkiss:</picture>

File: packages/edge/src/middleware-helpers.ts (hunk 2 above)
Comment (molebox) [reply packages/edge/src/middleware-helpers.ts]: lgtm

**Pull request merged**